msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JsonHelper.java

```diff
@@ -40,15 +41,77 @@ static <T> T convertJsonToObject(final String json, final Class<T> tClass) {
     }
 
     static IdToken createIdTokenFromEncodedTokenString(String token) {
-        String idTokenJson;
+        return JsonHelper.convertJsonToObject(getTokenPayloadClaims(token), IdToken.class);
+    }
+
+    static String getTokenPayloadClaims(String token) {
         try {
-            idTokenJson = new String(Base64.getUrlDecoder().decode(token.split("\\.")[1]), StandardCharsets.UTF_8);
+            return new String(Base64.getUrlDecoder().decode(token.split("\\.")[1]), StandardCharsets.UTF_8);
         } catch (ArrayIndexOutOfBoundsException e) {
             throw new MsalClientException("Error parsing ID token, missing payload section.",
                     AuthenticationErrorCode.INVALID_JWT);
         }
+    }
+
+    //Converts a generic JSON string to a Map<String, Object> with relevant types
+    static Map<String, Object> parseJsonToMap(String jsonString) {
```

Comment:
Did you try `JsonReader.readMap(ReadValueCallback<JsonReader, T> valueReaderFunc)`? May be able to simplify your logic here some.

Reply:
Unfortunately, JsonReader.readMap does not seem to produce the same Map as the previous nimbus-based methods. For example, nimbus converted unix timestamps to Date objects, and some fields that could be arrays or single strings were always treated as arrays. So for backwards compatibility we need to match the old nimbus behavior.

In the latest commit I added new tests and a new handleSpecialFields() method to JsonHelper to correctly handle and better demonstrate this behavior.

Reply:
Fair enough. You could maybe save some headache around managing state (keeping track of if you have progressed the parser, etc) by just having a callback that handled those special cases, but this is fine.